segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMapper.java

```diff
@@ -180,14 +177,88 @@ public StreamSegmentMapper(ContainerMetadata containerMetadata, OperationLog dur
+                .thenComposeAsync(si ->
+                                // Need to create the state file before we throw any further exceptions in order to recover from
+                                // previous partial executions (where we created a segment but no or empty state file).
+                                this.stateStore.put(segmentName, getState(si, attributes), timer.getRemaining())
+                                               .thenRun(() -> {
+                                                   if (si.getLength() > 0) {
+                                                       throw new CompletionException(new StreamSegmentExistsException(segmentName));
+                                                   }
+                                               }),
+                        this.executor);
+    }
+
+    /**
+     * Exception handler in the case when a Segment/Transaction fails to be created in Storage. This only handles
+     * StreamSegmentExistsException; all other exceptions are "bubbled" up automatically.
+     * <p>
+     * This method simply checks the integrity of the State File; if it exists and is valid, then the Segment is considered
+     * fully created and the original exception is bubbled up. If the State File does not exist or it is not valid, then
+     * the most up-to-date information about the segment is returned (as it exists in Storage).
+     *
+     * @param segmentName       The name of the Segment/Transaction involved.
+     * @param originalException The exception that triggered this.
+     * @param timer             A TimeoutTimer to determine how much time is left to complete the operation.
+     * @return A CompletableFuture that, when completed normally, will contain information about the Segment. If the Segment
+     * already exists or another error happened, this will be completed with the appropriate exception.
+     */
+    private CompletableFuture<SegmentProperties> handleStorageCreateException(String segmentName, Throwable originalException, TimeoutTimer timer) {
+        if (!(originalException instanceof StreamSegmentExistsException)) {
+            // Some other kind of exception that we can't handle here.
+            return FutureHelpers.failedFuture(originalException);
+        }
+
+        return this.stateStore
+                .get(segmentName, timer.getRemaining())
+                .exceptionally(ex -> {
+                    // Segment exists, but the State File is missing or corrupt. We have the data needed to rebuild it,
+                    // so ignore any exceptions coming this way.
+                    log.warn("{}: Missing or corrupt State File for existing Segment '{}'; recreating.", this.traceObjectId, segmentName, ex);
+                    return null;
+                })
+                .thenComposeAsync(s -> {
+                    if (s == null) {
```

Comment:
I feel like this null is overloaded because it does not differentiate between stateStore.get() throwing vs returning null. 
Perhaps the code would be more clear if this thenComposeAsync and the exceptionally block above were combined into a single handleAsync block

Reply:
Well, I intentionally coalesce errors from `get` and it actually returning `null` into `null` because, as the description indicates, we treat missing and invalid state files as the same.

Reply:
It is not obvious to me that this is correct for some arbitrary throwable. Shouldn't we handle errors differently from the special case of it being empty that we are handling.

Reply:
Fixed to only ignore `StreamSegmentNotExistsException` and `DataCorruptionException`.